### PR TITLE
[Spike] Accept offers unconditionally

### PR DESCRIPTION
**Pull request by steventux**
## Context

We need to understand the technical complexities of making unconditional offers.

## Changes proposed in this pull request

I found that you can legitimately make an offer via the API and not include conditions. No validation seems to prevent this.
So this spike adds a way for a candidate to accept an offer with no conditions, transitioning it from `offered` to `recruited` skipping any conditions related states and provider workflow.

## Guidance to review

To test:

1. Using a 'Received' application make an offer via curl, Postman with data like:
```
{
  "data": {
    "conditions": [],
    "course": null
  },
  "meta": {
    "attribution": {
      "full_name": "Jane Smith",
      "email": "jane.smith@example.com",
      "user_id": "12345"
    },
    "timestamp": "2019-03-01T15:33:49.216Z"
  }
}
```

2. Check the application (this can be done in the provider UI) now has an offer with no conditions.
3. Sign in to support and find the same application, login as the applying candidate
4. Go to the application offer tab and accept the offer

## TODO:

`AcceptUnconditionalOffer` Service spec
CandidateMailer method/template for unconditional offers.

## Link to Trello card

https://trello.com/c/uNuG9vGi/3397-tech-spike-supporting-unconditional-offers-via-the-api

## Things to check

- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/docs/environment-variables.md#azure-hosting-devops-pipeline)

### CODE REVIEW

**Review by duncanjbrown — COMMENTED** [review]
Nice and simple! I wonder if we want to start the `if` branch earlier and accept a bit of duplication (which we could then factor out) between conditional and unconditional acceptance? eg the emails probably won't be the same, we should probably have a dedicated state transition from offer -> recruited etc

**Comment by steventux** [comment]
@duncanjbrown This draft PR should be up to date with your suggestions, it should be fairly simple to use this as a basis for enabling this feature.

File: app/services/accept_unconditional_offer.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  end
+
+  def save!
+    ActiveRecord::Base.transaction do
+      ApplicationStateChange.new(@application_choice).accept_unconditional_offer!
+      @application_choice.update!(accepted_at: Time.zone.now, recruited_at: Time.zone.now)
+    end
+
+    # TODO: Mailer method for unconditional offers.
+    CandidateMailer.offer_accepted(@application_choice).deliver_later
```
Comment (duncanjbrown) [inline app/services/accept_unconditional_offer.rb:13]: Should we have a provider email here too?

**Comment by steventux** [comment]
Closing as we're implementing here https://github.com/DFE-Digital/apply-for-teacher-training/pull/4154

**Pull request closed without merging**